Trying to send an empty message results in an error

Sending an empty message results in this error:

```
org.springframework.jms.MessageEOFException: Cannot read all of non-RMQ Message body.; nested exception is javax.jms.MessageEOFException: Cannot read all of non-RMQ Message body.
	at org.springframework.jms.support.JmsUtils.convertJmsAccessException(JmsUtils.java:294)
	at org.springframework.jms.support.JmsAccessor.convertJmsAccessException(JmsAccessor.java:169)
	at org.springframework.jms.core.JmsTemplate.execute(JmsTemplate.java:487)
	at org.apache.camel.component.jms.JmsConfiguration$CamelJmsTemplate.send(JmsConfiguration.java:484)
	at org.apache.camel.component.jms.JmsProducer.doSend(JmsProducer.java:440)
	at org.apache.camel.component.jms.JmsProducer.processInOnly(JmsProducer.java:394)
	at org.apache.camel.component.jms.JmsProducer.process(JmsProducer.java:157)
	at org.apache.camel.processor.SendProcessor.process(SendProcessor.java:145)
	at org.apache.camel.management.InstrumentationProcessor.process(InstrumentationProcessor.java:77)
	at org.apache.camel.processor.RedeliveryErrorHandler.process(RedeliveryErrorHandler.java:541)
	at org.apache.camel.processor.CamelInternalProcessor.process(CamelInternalProcessor.java:198)
	at org.apache.camel.processor.Pipeline.process(Pipeline.java:120)
	at org.apache.camel.processor.Pipeline.process(Pipeline.java:83)
	at org.apache.camel.processor.CamelInternalProcessor.process(CamelInternalProcessor.java:198)
	at org.apache.camel.component.netty4.handlers.ServerChannelHandler.processAsynchronously(ServerChannelHandler.java:138)
	at org.apache.camel.component.netty4.handlers.ServerChannelHandler.channelRead0(ServerChannelHandler.java:109)
	at org.apache.camel.component.netty4.http.handlers.HttpServerChannelHandler.channelRead0(HttpServerChannelHandler.java:213)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105)
	at org.apache.camel.component.netty4.http.handlers.HttpServerMultiplexChannelHandler.channelRead0(HttpServerMultiplexChannelHandler.java:113)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.access$600(AbstractChannelHandlerContext.java:38)
	at io.netty.channel.AbstractChannelHandlerContext$7.run(AbstractChannelHandlerContext.java:353)
	at io.netty.util.concurrent.DefaultEventExecutor.run(DefaultEventExecutor.java:66)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at java.lang.Thread.run(Thread.java:745)
Caused by: javax.jms.MessageEOFException: Cannot read all of non-RMQ Message body.
	at com.rabbitmq.jms.client.message.RMQBytesMessage.recreate(RMQBytesMessage.java:561)
	at com.rabbitmq.jms.client.RMQMessage.normalise(RMQMessage.java:1356)
	at com.rabbitmq.jms.client.RMQMessageProducer.internalSend(RMQMessageProducer.java:234)
	at com.rabbitmq.jms.client.RMQMessageProducer.send(RMQMessageProducer.java:189)
	at org.springframework.jms.core.JmsTemplate.doSend(JmsTemplate.java:626)
	at org.apache.camel.component.jms.JmsConfiguration$CamelJmsTemplate.doSend(JmsConfiguration.java:592)
	at org.apache.camel.component.jms.JmsConfiguration$CamelJmsTemplate.doSendToDestination(JmsConfiguration.java:531)
	at org.apache.camel.component.jms.JmsConfiguration$CamelJmsTemplate.access$100(JmsConfiguration.java:473)
	at org.apache.camel.component.jms.JmsConfiguration$CamelJmsTemplate$1.doInJms(JmsConfiguration.java:487)
	at org.springframework.jms.core.JmsTemplate.execute(JmsTemplate.java:484)
	... 23 more
```

It seems that the problem is in `RMQBytesMessage#recreate(BytesMessage)`, line `if (bodySize != msg.readBytes(byteArray, bodySize))`. If the stream is empty, readBytes returns `-1`, which corresponds to bodySize being `0`. However with the current straightforward comparison `-1` and `0` are not equal.

--- a/src/main/java/com/rabbitmq/jms/client/RMQMessage.java
+++ b/src/main/java/com/rabbitmq/jms/client/RMQMessage.java
@@ -1360,7 +1360,7 @@ public abstract class RMQMessage implements Message, Cloneable {
         return dft;
     }
 
-    static RMQMessage normalise(Message msg, List<String> trustedPackages) throws JMSException {
+    static RMQMessage normalise(Message msg) throws JMSException {
         if (msg instanceof RMQMessage) return (RMQMessage) msg;
 
         /* If not one of our own, copy it into an RMQMessage */

--- a/src/main/java/com/rabbitmq/jms/client/RMQMessageProducer.java
+++ b/src/main/java/com/rabbitmq/jms/client/RMQMessageProducer.java
@@ -2,7 +2,6 @@
 package com.rabbitmq.jms.client;
 
 import java.io.IOException;
-import java.util.List;
 
 import javax.jms.Destination;
 import javax.jms.InvalidDestinationException;
@@ -63,7 +62,6 @@ public class RMQMessageProducer implements MessageProducer, QueueSender, TopicPu
      * The default TTL value is 0 (zero), meaning indefinite (no time-out).
      */
     private long ttl = Message.DEFAULT_TIME_TO_LIVE;
-    private List<String> trustedPackages;
 
     /**
      * Create a producer of messages.
@@ -73,8 +71,6 @@ public class RMQMessageProducer implements MessageProducer, QueueSender, TopicPu
     public RMQMessageProducer(RMQSession session, RMQDestination destination) {
         this.session = session;
         this.destination = destination;
-
-        this.trustedPackages = session.getTrustedPackages();
     }
 
     /**
@@ -231,7 +227,7 @@ public class RMQMessageProducer implements MessageProducer, QueueSender, TopicPu
             deliveryMode = javax.jms.DeliveryMode.NON_PERSISTENT;
 
         /* Normalise message to internal form */
-        RMQMessage rmqMessage = RMQMessage.normalise(message, this.trustedPackages);
+        RMQMessage rmqMessage = RMQMessage.normalise(message);
 
         /* Set known JMS message properties that need to be set during this call */
         long currentTime = System.currentTimeMillis();

--- a/src/main/java/com/rabbitmq/jms/client/message/RMQBytesMessage.java
+++ b/src/main/java/com/rabbitmq/jms/client/message/RMQBytesMessage.java
@@ -557,8 +557,10 @@ public class RMQBytesMessage extends RMQMessage implements BytesMessage {
             RMQBytesMessage rmqBMsg = new RMQBytesMessage();
             RMQMessage.copyAttributes(rmqBMsg, msg);
             byte[] byteArray = new byte[bodySize];
-            if (bodySize != msg.readBytes(byteArray, bodySize)) // must read the whole body at once
-                throw new MessageEOFException("Cannot read all of non-RMQ Message body.");
+            if (bodyLength > 0) {
+                if (bodySize != msg.readBytes(byteArray, bodySize)) // must read the whole body at once
+                    throw new MessageEOFException("Cannot read all of non-RMQ Message body.");
+            }
             rmqBMsg.writeBytes(byteArray);
             rmqBMsg.reset();  // make body read-only and set pointer to start.
             return rmqBMsg;

--- a/src/test/java/com/rabbitmq/jms/client/ForeignBytesMessageTest.java
+++ b/src/test/java/com/rabbitmq/jms/client/ForeignBytesMessageTest.java
@@ -14,6 +14,8 @@ import javax.jms.Destination;
 import javax.jms.JMSException;
 import javax.jms.MessageFormatException;
 
+import com.rabbitmq.jms.client.message.RMQBytesMessage;
+import org.junit.Assert;
 import org.junit.Test;
 
 public class ForeignBytesMessageTest {
@@ -22,9 +24,17 @@ public class ForeignBytesMessageTest {
     public void testBytesMessage() throws Exception {
         BytesMessage testMsg = new ForeignBytesMessage();
         populateBytesMessage(testMsg);
-        BytesMessage rmqMsg = (BytesMessage) RMQMessage.normalise(testMsg, Collections.singletonList("*"));
+        BytesMessage rmqMsg = (BytesMessage) RMQMessage.normalise(testMsg);
         rmqMsg.reset();
         inspectBytesMessage(rmqMsg);
+    }
+
+    @Test
+    public void testEmptyBytesMessage() throws Exception {
+        BytesMessage testMsg = new RMQBytesMessage();
+        assertEquals(0, testMsg.getBodyLength());
+        BytesMessage rmqMsg = (BytesMessage) RMQBytesMessage.recreate(testMsg);
+        assertEquals(0, rmqMsg.getBodyLength());
     }
 
     private static final byte[] BYTE_ARRAY = new byte[]{ 0,1,2, (byte) -1 };

